### PR TITLE
feat: observability (api, metrics, and startup / liveliness probes)

### LINKED ISSUE #70: feat: observability
### Problem
As the standalone watch-tower moves into production, observability is critical in ensuring the reliability of the service.

### Suggested solution
Introduce an express API that provides health & readiness monitoring, in addition to a metrics collection endpoint for automatic prometheus scraping.

### Alternatives considered
Use of loggly / sentry. These have been trialed, and while loggly is suitable for some log analysis, this can easily be achieved by ELK. Loggly as well doesn't provide much in terms of metrics monitoring - this is the bread and butter of Prometheus / Grafana.

### Acceptance criteria

- [x] `/health` endpoint that returns not ready if the chain is still in warm-up.
- [x] `/metrics` endpoint with targeted metrics distributed throughout.

### PR DESCRIPTION
**Pull request by mfw78**
# Description
This PR adds observability to the watch-tower so that it can be monitored in production.

# Changes

- [x] Add express API with `/api/dump` endpoint
- [x] Add prometheus metrics at `/metrics`
- [ ] Add startup and liveliness probes at `/health`
- [ ] Add CLI options for enabling API (enabled by default)

## How to test

1. Start watch-tower.
2. Check above endpoints. For `dump`, add chain Id, for example: `/api/dump/1`

## Related Issues

Fixes #70

### CODE REVIEW

File: src/utils/api.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  private bootstrap() {
+    const collectDefaultMetrics = client.collectDefaultMetrics;
+    collectDefaultMetrics();
+
+    this.app.use(express.json());
+    this.app.use(express.urlencoded({ extended: true }));
+    this.app.get("/", (_req: Request, res: Response) => {
+      res.send("Moooo!");
```
Comment (anxolin) [inline src/utils/api.ts:28]: where is the emoji! 🐮

File: src/commands/run.ts [hunk 2]
```diff
@@ -1,23 +1,31 @@
...
 /**
  * Run the watch-tower 👀🐮
  * @param options Specified by the CLI / environment for running the watch-tower
  */
 export async function run(options: RunOptions) {
   const log = getLogger("commands:run");
-  const { rpc, deploymentBlock, oneShot } = options;
+  const { rpc, deploymentBlock, oneShot, enableApi, apiPort } = options;
+
+  const api = new ApiService(apiPort);
```
Comment (anxolin) [inline src/commands/run.ts:14]: can we conditionally create this api by option (default could be to create)

File: src/utils/api.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  get getApp(): Express {
+    return this.app;
+  }
+
+  async start(): Promise<Server> {
+    return await new Promise((resolve, reject) => {
+      try {
+        const log = getLogger("api");
```
Comment (anxolin) [inline src/utils/api.ts:46]: should we move this log to at the beginging of the file?

File: src/utils/api.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        this.server.close();
+      } catch (err) {
+        reject(err);
+      }
+    });
+  }
+}
+
+const dumpRoute = (router: Router) => {
+  router.get("/dump/:chainId", async (req: Request, res: Response) => {
```
Comment (anxolin) [inline src/utils/api.ts:81]: this is cool :)

File: src/utils/metrics.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
+import client from "prom-client";
+
+export class MetricsService {
```
Comment (anxolin) [inline src/utils/metrics.ts:3]: is there any added benefit from wrapping the metrics in a Class? I would just export them directly

**Review by anxolin — COMMENTED** [review]
First small review, for now i didn't check the metrics itself, i will do soon another pass. Just to give a quick feedback

File: src/utils/api.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+  async start(): Promise<Server> {
+    return await new Promise((resolve, reject) => {
+      try {
+        const log = getLogger("api");
+        if (this.server?.listening) {
+          throw new Error("Server is already running");
+        }
+        this.server = this.app.listen(this.PORT, () => {
+          log.info(`Starting Rest API server on port ${this.PORT}`);
```
Comment (anxolin) [inline src/utils/api.ts:51]: nit: can you add also the URL, normally is convenient to just click on the URL in the console to open a browser

**Comment by anxolin** [comment]
Super cool! works like a charm Will do now another run with more eyes on the metrics themselves

image

File: src/utils/metrics.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    help: "Total number of reorgs",
+    labelNames: ["chain_id"],
+  });
+  static blockWatcherBlockTime = new client.Gauge({
+    name: "block_watcher_block_time",
+    help: "Time since last block",
+    labelNames: ["chain_id"],
+  });
+  static blockWatcherNumEventsProcessed = new client.Counter({
+    name: "block_watcher_num_events_processed",
```
Comment (anxolin) [inline src/utils/metrics.ts:25]: I think a more prometheus like name is `watch_tower_events_total`

File: src/utils/metrics.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    help: "Total number of merkle root sets added",
+    labelNames: ["chain_id"],
+  });
+  static addNewOwnerErrorCount = new client.Counter({
+    name: "add_contract_error_count",
+    help: "Total number of errors adding contracts",
+    labelNames: ["chain_id", "function"],
+  });
+  static addContractRunDuration = new client.Histogram({
+    name: "add_contract_run_duration",
```
Comment (anxolin) [inline src/utils/metrics.ts:50]: You need to add the units `_duration_second` as per https://prometheus.io/docs/practices/naming/

**Review by anxolin — COMMENTED** [review]
I gave a bit of feedback in the metrics themselves. I need to go now, but i can do another review later. 

I'm not convinced with the naming, happy to PR on top. I just wrote some examples on some small improvements. 

Happy also to PR on top and apply some changes myself, but I don't want to get conflicts so I would rather wait for you to finish it

**Comment by anxolin** [comment]
@mfw78 BTW, this PR is already big, i wouldn't add the liveness probes here.

File: src/commands/run.ts (hunk 2 above)
Comment (mfw78) [reply src/commands/run.ts:14]: Done. Instead now it's specifically required to do `--disable-api`. The objects etc are only created in the event that it's not explicitly disabled.

File: src/utils/api.ts (hunk 3 above)
Comment (mfw78) [reply src/utils/api.ts:46]: This is not possible now due to the initialization requirements introduced with `initLogging`.

File: src/utils/metrics.ts (hunk 5 above)
Comment (mfw78) [reply src/utils/metrics.ts:3]: Nothing explicit unless we want to at some stage allow for metrics to be specified on a different port to the API, and therefore would have an object here that instantiates a `HttpServer`.

File: src/utils/api.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+  private bootstrap() {
+    const collectDefaultMetrics = client.collectDefaultMetrics;
+    collectDefaultMetrics();
+
+    this.app.use(express.json());
+    this.app.use(express.urlencoded({ extended: true }));
+    this.app.get("/", (_req: Request, res: Response) => {
+      res.send("Moooo!");
+      MetricsService.apiRequestCounter.labels("GET", "/", "200").inc();
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/utils/api.ts:29]: Done.

File: src/utils/metrics.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
+import client from "prom-client";
+
+export class MetricsService {
+  static apiRequestCounter = new client.Counter({
+    name: "api_request_count",
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/utils/metrics.ts:5]: Great reference on the naming, this makes absolutely perfect sense. Will adjust accordingly.

File: src/utils/metrics.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    help: "Total number of API requests",
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/utils/metrics.ts:6]: Done.

File: src/utils/metrics.ts (hunk 8 above)
Comment (mfw78) [reply src/utils/metrics.ts:50]: Have done so. Per the naming practices as well, have used the plural form of the `seconds`.

File: src/utils/metrics.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (mfw78) [reply src/utils/metrics.ts:5]: Done

File: src/utils/api.ts (hunk 3 above)
Comment (anxolin) [reply src/utils/api.ts:46]: yep, i see. not a big deal then

**Pull request merged**